components/movementsensor/replay/replay.go

```diff
@@ -0,0 +1,479 @@
+)
+
+var (
+	// model is the model of a replay movement sensor.
+	model = resource.DefaultModelFamily.WithModel("replay")
+
+	// ErrEndOfDataset represents that the replay sensor has reached the end of the dataset.
+	ErrEndOfDataset = errors.New("reached end of dataset")
+
+	// methodList is a list of all the base methods possible for a movement sensor to implement.
+	methodList = []string{"Position", "Orientation", "AngularVelocity", "LinearVelocity", "LinearAcceleration", "CompassHeading"}
+)
+
+func init() {
+	resource.RegisterComponent(movementsensor.API, model, resource.Registration[movementsensor.MovementSensor, *Config]{
+		Constructor: newReplayMovementSensor,
+	})
+}
+
+// Validate checks that the config attributes are valid for a replay movement sensor.
+func (cfg *Config) Validate(path string) ([]string, error) {
+	if cfg.Source == "" {
+		return nil, goutils.NewConfigValidationFieldRequiredError(path, "source")
+	}
+
+	var err error
+	var startTime time.Time
+	if cfg.Interval.Start != "" {
+		startTime, err = time.Parse(timeFormat, cfg.Interval.Start)
+		if err != nil {
+			return nil, errors.New("invalid time format for start time (UTC), use RFC3339")
+		}
+		if startTime.After(time.Now()) {
+			return nil, errors.New("invalid config, start time (UTC) must be in the past")
+		}
+	}
+
+	var endTime time.Time
+	if cfg.Interval.End != "" {
+		endTime, err = time.Parse(timeFormat, cfg.Interval.End)
+		if err != nil {
+			return nil, errors.New("invalid time format for end time (UTC), use RFC3339")
+		}
+		if endTime.After(time.Now()) {
+			return nil, errors.New("invalid config, end time (UTC) must be in the past")
+		}
+	}
+
+	if cfg.Interval.Start != "" && cfg.Interval.End != "" && startTime.After(endTime) {
+		return nil, errors.New("invalid config, end time (UTC) must be after start time (UTC)")
+	}
+
+	if cfg.BatchSize != nil && (*cfg.BatchSize > uint64(maxCacheSize) || *cfg.BatchSize == 0) {
+		return nil, errors.Errorf("batch_size must be between 1 and %d", maxCacheSize)
+	}
+
+	return []string{cloud.InternalServiceName.String()}, nil
+}
+
+// Config describes how to configure the replay movement sensor.
+type Config struct {
+	Source         string          `json:"source,omitempty"`
+	RobotID        string          `json:"robot_id,omitempty"`
+	LocationID     string          `json:"location_id,omitempty"`
+	OrganizationID string          `json:"organization_id,omitempty"`
+	Interval       TimeInterval    `json:"time_interval,omitempty"`
+	BatchSize      *uint64         `json:"batch_size,omitempty"`
+	Properties     map[string]bool `json:"properties,omitempty"`
+}
+
+// TimeInterval holds the start and end time used to filter data.
+type TimeInterval struct {
+	Start string `json:"start,omitempty"`
+	End   string `json:"end,omitempty"`
+}
+
+// cacheEntry stores data that was downloaded from a previous operation but has not yet been passed
+// to the caller.
+type cacheEntry struct {
+	data          *structpb.Struct
```

Comment:
note to self, we don't need an id here because for tabular data, we can download all the data at once rather than download the metadata and fill in the data